src/coreclr/jit/objectalloc.cpp

```diff
@@ -1334,3 +1482,1724 @@ void ObjectAllocator::RewriteUses()
+            if (isReachableOnGDVSuccess)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by failing GDV\n");
+            continue;
+        }
+        else
+        {
+            bool const isReachableOnGDVFailure = (trueSuccessorDominates && guardingRelop->OperIs(GT_NE)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_EQ));
+            if (isReachableOnGDVFailure)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by successful GDV\n");
+            continue;
+        }
+    }
+
+    JITDUMP("... no more doms\n");
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// IsGuard: does block look like a GDV guard
+//
+// Arguments:
+//   block -- block in question
+//   info -- [out] guard info
+//
+// Returns:
+//   Comparison tree if this is a guard, or nullptr
+//
+GenTree* ObjectAllocator::IsGuard(BasicBlock* block, GuardInfo* info)
+{
+    if (!block->KindIs(BBJ_COND))
+    {
+        JITDUMP("... not BBJ_COND\n");
+        return nullptr;
+    }
+
+    Statement* const stmt = block->lastStmt();
+    if (stmt == nullptr)
+    {
+        JITDUMP("... no stmt\n");
+        return nullptr;
+    }
+
+    GenTree* const jumpTree = stmt->GetRootNode();
+    if (!jumpTree->OperIs(GT_JTRUE))
+    {
+        JITDUMP("... no JTRUE\n");
+        return nullptr;
+    }
+
+    GenTree* const tree = jumpTree->AsOp()->gtOp1;
+
+    // Must be an equality or inequality
+    //
+    if (!tree->OperIs(GT_NE, GT_EQ))
+    {
+        JITDUMP("... not NE/EQ\n");
+        return nullptr;
+    }
+
+    GenTree* op1     = tree->AsOp()->gtOp1;
+    GenTree* op2     = tree->AsOp()->gtOp2;
+    bool     swapped = false;
+
+    // gdv creates NE(hnd, indir(locl))
+    // but let's not rely on that
+    //
+    if (!op1->OperIs(GT_IND))
+    {
+        swapped = true;
+        std::swap(op1, op2);
+    }
+
+    if (!op1->OperIs(GT_IND))
+    {
+        JITDUMP("... no JTRUE(cmp(ind, ...))\n");
+        return nullptr;
+    }
+
+    if (!op1->TypeIs(TYP_I_IMPL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int, ...))\n");
+        return nullptr;
+    }
+
+    GenTree* const addr = op1->AsIndir()->Addr();
+
+    if (!addr->TypeIs(TYP_REF))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(*:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!addr->OperIs(GT_LCL_VAR))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!op2->IsIconHandle(GTF_ICON_CLASS_HDL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), clsHnd))\n");
+        return nullptr;
+    }
+
+    // Passed the checks... fill in the info.
```

Comment:
Does this function care if the condition is EQ or NE?

Reply:
Not this one, no. `IsGuarded` checks though.